src/grappler/ngraph_optimizer.cc

```diff
@@ -0,0 +1,188 @@
+  opts.expect_device_spec = true;
+  Graph graph(OpRegistry::Global());
+  TF_RETURN_IF_ERROR(ConvertGraphDefToGraph(opts, item.graph, &graph));
+
+  // Get the nodes to be skipped
+  std::vector<string> skip_these_nodes;
+  for (const string& f : item.fetch) {
+    NGRAPH_VLOG(5) << "Skip fetch node: " << f;
+    int pos = f.find(":");
+    skip_these_nodes.push_back(f.substr(0, pos));
+  }
+
+  //
+  // Variable capture: Part that replaces all instances of VariableV2 with the
+  // NGraphVariable op. Making this replacement allows us to substitute in a
+  // kernel that tracks the freshness of variables (invalidating freshness when
+  // the reference is handed off to an "untrusted" op).
+  //
+
+  // For filename generation purposes, grab a fresh index. This is just an
+  // arbitrary integer to avoid filename collisions resulting from subsequent
+  // runs of this pass.
+  int idx = FreshIndex();
+
+  // If requested, dump pre-capture graphs.
+  if (DumpPrecaptureGraphs()) {
+    DumpGraphs(graph, idx, "precapture", "Pre-Capture Graph");
+  }
+
+  // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
+  // we will not do anything; all subsequent
+  // passes become a no-op.
+  if (config::IsEnabled() == false ||
+      std::getenv("NGRAPH_TF_DISABLE") != nullptr) {
+    return Status::OK();
+  }
+
+  // Do variable capture then, if requested, dump the graphs.
+  TF_RETURN_IF_ERROR(CaptureVariables(&graph, skip_these_nodes));
+  if (DumpCapturedGraphs()) {
+    DumpGraphs(graph, idx, "captured", "Graph With Variables Captured");
+  }
+
+  //
+  // Encapsulation: Part that rewrites the graph for nGraph operation.
+  //
+  // The part has several phases, each executed in sequence:
+  //
+  //   1. Marking [ngraph_mark_for_clustering.cc]
+  //   2. Cluster Assignment [ngraph_assign_clusters.cc]
+  //   3. Cluster Deassignment [ngraph_deassign_clusters.cc]
+  //   4. Cluster Encapsulation [ngraph_encapsulate_clusters.cc] - currently
+  //      part of the ngraph_rewrite_pass.cc to be executed after POST_REWRITE
+  //
+  // Between phases, graph dumps (in both .dot and .pbtxt format) may be
+  // requested by setting the following environment variables:
+  //
+  //   NGRAPH_TF_DUMP_UNMARKED_GRAPHS=1      dumps graphs before phase 1
+  //   NGRAPH_TF_DUMP_MARKED_GRAPHS=1        dumps graphs after phase 1
+  //   NGRAPH_TF_DUMP_CLUSTERED_GRAPHS=1     dumps graphs after phase 2
+  //   NGRAPH_TF_DUMP_DECLUSTERED_GRAPHS=1   dumps graphs after phase 3
+  //   NGRAPH_TF_DUMP_ENCAPSULATED_GRAPHS=1  dumps graphs after phase 4
+  //   NGRAPH_TF_DUMP_GRAPHS=1               all of the above
+  //
+
+  // If requested, dump unmarked graphs.
+  if (DumpUnmarkedGraphs()) {
+    DumpGraphs(graph, idx, "unmarked", "Unmarked Graph");
+  }
+
+  // If ngraph is disabled via ngraph_bridge api or NGRAPH_TF_DISABLE is set
+  // we will not do anything; all subsequent
+  // passes become a no-op.
+  if (config::IsEnabled() == false ||
```

Comment:
We dont need this check as we are exiting after dumping precaptured.

Reply:
Done. Thanks for catching this.